### PR TITLE
Add basic DFU transfer support.

### PR DESCRIPTION
**Pull request by nomis52**
Right now the data is just discarded. The next step will be to write to the
flash.

### CODE REVIEW

File: Bootloader/firmware/src/app.c [hunk 1]
```diff
@@ -18,45 +18,206 @@
...
+    StallAndError(DFU_STATUS_ERR_STALLED_PKT);
+    return;
+  }
+
+  if (packet->wLength) {
+    USB_DEVICE_ControlReceive(g_app.usb_device, DATA_BUFFER, packet->wLength);
   } else {
-    USB_DEVICE_ControlStatus(g_app.usb_device,
-                             USB_DEVICE_CONTROL_STATUS_ERROR);
+    // Verify image here
```
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: TODO

File: Bootloader/firmware/src/app.c [hunk 2]
```diff
@@ -18,45 +18,206 @@
...
+    g_app.dfu_state = DFU_STATE_MANIFEST_SYNC;
+    g_app.has_new_firmware = true;
+    USB_DEVICE_ControlStatus(g_app.usb_device, USB_DEVICE_CONTROL_STATUS_OK);
+  }
+}
+
+static inline void DFUGetStatus() {
+  // Some Get Status messages trigger a state change.
+  // The status resposne always contains the *next* state, so figure that out
```
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: SPaG response

File: Bootloader/firmware/src/app.c [hunk 3]
```diff
@@ -101,33 +262,37 @@ static void USBEventHandler(USB_DEVICE_EVENT event,
...
         USB_DEVICE_ControlStatus(g_app.usb_device,
                                  USB_DEVICE_CONTROL_STATUS_ERROR);
       }
       break;
 
+    case USB_DEVICE_EVENT_CONTROL_TRANSFER_DATA_RECEIVED:
+      DFUTransferComplete();
+      break;
+    case USB_DEVICE_EVENT_CONTROL_TRANSFER_DATA_SENT:
+      // The harmony examples show a call to USB_DEVICE_ControlStatus() here,
```
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: SPaG Harmony

File: Bootloader/firmware/src/app.c [hunk 4]
```diff
@@ -136,52 +301,56 @@ static void USBEventHandler(USB_DEVICE_EVENT event,
...
+      if (g_app.dfu_state == DFU_STATE_MANIFEST) {
+        g_app.has_new_firmware = false;
+        // pretend we're done and switch back to DFU_STATE_MANIFEST_SYNC;
+        g_app.dfu_state = DFU_STATE_MANIFEST_SYNC;
       }
       break;
-    case APP_STATE_ERROR:
-      break;
+    case APP_STATE_BOOT:
+      // Do boot
```
Comment (peternewman) [inline Bootloader/firmware/src/app.c]: TODO

**Comment by peternewman** [comment]
Generally LGTM

File: Bootloader/firmware/src/app.c (hunk 1 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.c]: Done

File: Bootloader/firmware/src/app.c (hunk 2 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.c]: Fixed

File: Bootloader/firmware/src/app.c (hunk 3 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.c]: Fixed

File: Bootloader/firmware/src/app.c (hunk 4 above)
Comment (nomis52) [inline Bootloader/firmware/src/app.c]: Done

**Comment by peternewman** [comment]
LGTm

**Pull request merged**